src/consensus/dkg.rs

```diff
@@ -0,0 +1,223 @@
+        let mut broadcast = Vec::new();
+        for (key, key_gen) in self.key_gen_map.iter_mut() {
+            debug!("Progressing DKG {:?}", key);
+            if let Ok(messages) = key_gen.timed_phase_transition(rng) {
+                for message in messages {
+                    broadcast.push((key.clone(), message));
+                }
+            }
+        }
+
+        broadcast
+    }
+
+    pub fn process_dkg_message(
+        &mut self,
+        rng: &mut MainRng,
+        dkg_key: &DkgKey,
+        message: DkgMessage<PublicId>,
+    ) -> Vec<DkgMessage<PublicId>> {
+        let mut messages = Vec::new();
+
+        if let Some(mut key_gen) = self.key_gen_map.remove(dkg_key) {
+            if let Ok(responses) = key_gen.handle_message(rng, message) {
+                messages = responses;
+            }
+
+            let _ = self.key_gen_map.insert(dkg_key.clone(), key_gen);
+        }
+        messages
+    }
+
+    pub fn init_dkg_gen(
+        &mut self,
+        full_id: &FullId,
+        dkg_key: &DkgKey,
+    ) -> Vec<DkgMessage<PublicId>> {
+        if self.key_gen_map.contains_key(dkg_key) || self.completed_dkg.contains(dkg_key) {
+            trace!("already have key_gen of {:?}", dkg_key);
+            return vec![];
+        }
+
+        let threshold = threshold_count(dkg_key.0.len());
+
+        if let Ok((key_gen, message)) = KeyGen::initialize(full_id, threshold, dkg_key.0.clone()) {
+            debug!("started key_gen of {:?}", dkg_key);
+
+            let _ = self.key_gen_map.insert(dkg_key.clone(), key_gen);
+            vec![message]
+        } else {
+            vec![]
+        }
+    }
+
+    pub fn has_info(&mut self, dkg_key: &DkgKey, dkg_result: &DkgResult) -> bool {
+        if self.dkg_cache.contains_key(dkg_key) {
+            true
+        } else {
+            // During split, sibling DKG got completed before this elder notice the churn.
+            // In this case, as this elder is responsible for voting SectionInfo and Key for the
+            // sibling DKG, the notified result has to be cached.
+            if !self.completed_dkg.contains(dkg_key) {
+                let _ = self
+                    .sibling_dkg_result_cache
+                    .insert(dkg_key.clone(), dkg_result.clone());
+            }
+            false
+        }
+    }
+
+    pub fn push_info(&mut self, dkg_key: &DkgKey, info: EldersInfo) -> Option<DkgResult> {
+        let _ = self.dkg_cache.insert(dkg_key.clone(), info);
+
+        if let Some(dkg_result) = self.sibling_dkg_result_cache.remove(dkg_key) {
+            Some(dkg_result)
+        } else {
+            None
+        }
+    }
+
+    pub fn take_info(&mut self, dkg_key: &DkgKey) -> Option<EldersInfo> {
+        self.dkg_cache.remove(dkg_key)
+    }
+
+    pub fn info_keys(&self) -> impl Iterator<Item = &DkgKey> {
+        self.dkg_cache.keys()
+    }
+
+    pub fn insert_dkg_result(&mut self, dkg_key: DkgKey, dkg_result: DkgResult) {
```

Comment:
Maybe better to rename this to `insert_sibling_dkg_result` or something like that?

Reply:
ok, I will change to `insert_sibling_dkg_result`